src/QueryFunction.ts

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2020 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ElementHandle} from './JSHandle';
+
+export interface QueryFunction {
+  (element: Element, selector: string): (ElementHandle | null) | ElementHandle[];
+}
+
+const _customQueryFunctions = new Map<string, QueryFunction>();
+
+export function registerCustomQueryFunction(name: string, queryFunction: Function): void {
+  if (_customQueryFunctions.get(name))
```

Comment:
TODO: validate the name here. At a minimum, we should ensure it doesn't contain `SEARCH_SEPARATOR`. But IMHO we should be even more restrictive and only allow `/^[a-zA-Z]+$/`.

Reply:
👍 and then let's add some unit tests to check the behaviour

Reply:
Sure

Reply:
Done